### PR TITLE
Refactor input forms validation logic

### PR DESCRIPTION
**Pull request by balinskia**
Now we have reusable input form validators.
We have validators for:
* non negative float number
* non negative integer number

Changed absolute import path '@/...' to relative because of problems
with importing in tests - '@/...' is a vue import and in test we don't have it.

### CODE REVIEW

File: frontend/src/api/invest.ts [hunk 1]
```diff
@@ -1,5 +1,5 @@
 import API, { BigNumber, LoanState } from './index';
-import { LoanToInvestT } from '@/store/invest/types';
+import { LoanToInvestT } from '../store/invest/types';
```
Comment (q3k) [inline frontend/src/api/invest.ts]: Why the change to a relative path?

File: frontend/src/store/account/actions.ts [hunk 2]
```diff
@@ -13,4 +13,12 @@ const actions = {
   }
 }
 
+export function getMyBalance(): Promise<number> {
+  return new Promise(function (resolve, reject) {
+    api.getMyBalance(({ balance }: { balance: string }) => {
```
Comment (q3k) [inline frontend/src/store/account/actions.ts]: Does this function ever return an error? The promise will resolve but never reject.

File: frontend/src/api/api.ts [hunk 3]
```diff
@@ -12,17 +12,25 @@ let waiters: Array<Waiter> = [];
...
 export default {
   init: async function (events: EventsProvider): Promise<void> {
-    try {
-      api = new Client(METABACKEND.URL, METABACKEND.NETWORK);
-      await api.initialize();
-      events.afterSuccessfulInitialization();
-      // Wake up sleepers waiting for API.
-      for (let i = 0; i < waiters.length; i++) {
-        waiters[i](api);
+    async function initClient({ onSuccess }) {
```
Comment (q3k) [inline frontend/src/api/api.ts]: I don't get it. Why a callback here? It's an async function.

Comment (q3k) [reply frontend/src/api/api.ts]: Make it return a bool, for success or lack thereof.

File: frontend/src/api/api.ts [hunk 4]
```diff
@@ -12,17 +12,25 @@ let waiters: Array<Waiter> = [];
...
+        onSuccess();
+      } catch (error) {
+        handleInitError(error, events);
       }
-    } catch (error) {
-      handleInitError(error, events);
     }
+    const interval = setInterval(function () {
+      initClient({
+        onSuccess: function () { clearInterval(interval); }
```
Comment (q3k) [inline frontend/src/api/api.ts]: And here, instead of getting called back, check the return bool for success or not.

If you want to keep using setInterval (which isn't async), you can do

```typescript
initClient.then((success)=>{
    if (success) {
       clearInterval(interval)
    }
});
```

You can also find async-friendly setInterval implementations on npm.

File: frontend/src/api/invest.ts [hunk 5]
```diff
@@ -30,23 +30,43 @@ export async function getLoansToInvest(cb: (loans: LoanToInvestT[]) => void): Pr
   cb(loansToInvest);
 }
 
-export async function getLoanToInvest(shortId) {
+export async function getLoanToInvest(shortId: string, cb: (loan: LoanToInvestT) => void) {
```
Comment (q3k) [inline frontend/src/api/invest.ts]: Again, why the callback? Just return the value, it's an async function.

File: frontend/src/api/invest.ts [hunk 6]
```diff
@@ -30,23 +30,43 @@ export async function getLoansToInvest(cb: (loans: LoanToInvestT[]) => void): Pr
...
+
+export async function investInLoan(shortId: string, amount: number) {
+  const api = await API.instance();
+  const testToken = api.testToken;
+  let currentUser = await api.currentUser();
+  let loan = await api.loan(shortId);
+  await loan.invest(await testToken.integerize(new BigNumber(amount)));
+}
+
+export async function getMyInvestments(cb: (loans: LoanToInvestT[]) => void) {
```
Comment (q3k) [inline frontend/src/api/invest.ts]: Same, callback in an async function.

File: frontend/src/api/invest.ts [hunk 7]
```diff
@@ -30,23 +30,43 @@ export async function getLoansToInvest(cb: (loans: LoanToInvestT[]) => void): Pr
...
+  const api = await API.instance();
+  let user = await api.currentUser();
+  let blockchainLoans = await api.loansByOwner(user);
+  await Promise.all(blockchainLoans.map(loan => loan.updateStateFromBlockchain()));
+  cb(await Promise.all(blockchainLoans.map(blockchainLoanToVueLoan)));
+}
 
+async function blockchainLoanToVueLoan(blockchainLoan): Promise<LoanToInvestT> {
```
Comment (q3k) [inline frontend/src/api/invest.ts]: Add a type declaration to the argument.

Also, it's not really a blockchain loan, it's something more like Library Loan? And it's not a VueLoan, It's a ViewLoan. I think.

File: frontend/src/store/my-loans/actions.ts [hunk 8]
```diff
@@ -10,6 +13,21 @@ const actions = {
         loans
       })
     })
+  },
+  [TRANSFER_COLLATERAL]({ commit }, payload: { shortId: string, amount: string, onSuccess: () => void }) {
+    (async function () {
+      const { shortId, amount, onSuccess } = payload;
+      commit('START_TRANSFERING_COLLATERAL', { shortId });
+      const oldBalance: number = await getMyBalance();
+      await gatherCollateral(shortId, amount);
```
Comment (q3k) [inline frontend/src/store/my-loans/actions.ts]: Your error checking here won't work. If transferring the collateral fails, this await will throw an Error. You should probably do something like:

```typescript
gatherCollateral(shortId, amount).then(()=>{
   commit('COLLATERAL_TRANSFERED', { shortId });
   onSuccess();
}).catch((err: Error)=>{
  console.error(err);
  commit('SOME_GENERIC_ERROR_PAGE', { msg: "Transferring collateral failed. Do you have enough funds?" });
});
```

File: frontend/src/store/my-loans/mutations.ts [hunk 9]
```diff
@@ -9,4 +9,30 @@ export const mutations = {
...
+        return {
+          ...loan,
+          isTransferingCollateral: true
+        }
+      } else {
+        return loan;
+      }
+    })
+  },
+  'COLLATERAL_TRANSFERED': function (state: MyLoansStateT, { shortId }): void {
```
Comment (q3k) [inline frontend/src/store/my-loans/mutations.ts]: This mutation should maybe receive an updated loan from the library, and replace it fully instead? I'm not sure.

File: frontend/src/components/invest/MyInvestments/MyInvestmentTile.vue [hunk 10]
```diff
@@ -3,23 +3,22 @@
...
+        <div class="mit-short-desc">Demo Loan {{ investment.description }}</div>
         <div class="mit-user">{{ investment.userName }}</div>
       </div>
     </div>
     <div class="mit-financial-data">
       <div>
-        <div class="mit-fin-line-a"> CURRENT RATE </div>
-        <div class="mit-fin-line-b">{{ investment.currentRate }}</div>
+        <div class="mit-fin-line-a"> <div> CURRENT </div> <div> RATE </div> </div>
+        <div class="mit-fin-line-b">{{ investment.interestPermil }} %</div>
```
Comment (q3k) [inline frontend/src/components/invest/MyInvestments/MyInvestmentTile.vue]: Hm, is it a permil, or is it a percent? :)

File: frontend/src/api/invest.ts (hunk 5 above)
Comment (balinskia) [reply frontend/src/api/invest.ts]: right!

File: frontend/src/api/invest.ts (hunk 7 above)
Comment (balinskia) [reply frontend/src/api/invest.ts]: right!

File: frontend/src/api/invest.ts (hunk 6 above)
Comment (balinskia) [reply frontend/src/api/invest.ts]: right!

File: frontend/src/api/api.ts (hunk 4 above)
Comment (balinskia) [reply frontend/src/api/api.ts]: done!

File: frontend/src/store/my-loans/mutations.ts (hunk 9 above)
Comment (q3k) [reply frontend/src/store/my-loans/mutations.ts]: Ack.
